tapchannel/aux_sweeper.go

```diff
@@ -0,0 +1,1475 @@
+			InternalKey:    scriptTree.InternalKey,
+			NonAssetLeaves: sibling,
+			SortTaprootKeyBytes: schnorr.SerializePubKey(
+				scriptTree.TaprootKey,
+			),
+		})
+	}
+
+	localAnchor := anchorAlloc(keyRing.ToLocalKey)
+	remoteAnchor := anchorAlloc(keyRing.ToRemoteKey)
+
+	return lfn.AndThen2(
+		localAnchor, remoteAnchor,
+		func(a1, a2 *Allocation) lfn.Result[[]*Allocation] {
+			// Before we return the anchors, we'll make sure that
+			// they end up in the right sort order.
+			scriptCompare := bytes.Compare(
+				a1.SortTaprootKeyBytes, a2.SortTaprootKeyBytes,
+			)
+
+			if scriptCompare < 0 {
+				a1.OutputIndex = 0
+				a2.OutputIndex = 1
+			} else {
+				a2.OutputIndex = 0
+				a1.OutputIndex = 1
+			}
+
+			return lfn.Ok([]*Allocation{a1, a2})
+		},
+	)
+}
+
+// remoteCommitScriptKey creates the script key for the remote commitment
+// output.
+func remoteCommitScriptKey(remoteKey *btcec.PublicKey,
+) lfn.Result[asset.ScriptKey] {
+
+	remoteScriptTree, err := input.NewRemoteCommitScriptTree(
+		remoteKey, input.NoneTapLeaf(),
+	)
+	if err != nil {
+		return lfn.Err[asset.ScriptKey](err)
+	}
+
+	outputKey := asset.NewScriptKey(remoteScriptTree.TaprootKey).PubKey
+
+	return lfn.Ok(asset.ScriptKey{
+		PubKey: outputKey,
+		TweakedScriptKey: &asset.TweakedScriptKey{
+			RawKey: keychain.KeyDescriptor{
+				PubKey: remoteScriptTree.InternalKey,
+			},
+			Tweak: remoteScriptTree.TapscriptRoot,
+		},
+	})
+}
+
+// localCommitScriptKey creates the script key for the local commitment output.
+func localCommitScriptKey(localKey, revokeKey *btcec.PublicKey,
+	csvDelay uint32) lfn.Result[asset.ScriptKey] {
+
+	localScriptTree, err := input.NewLocalCommitScriptTree(
+		csvDelay, localKey, revokeKey, input.NoneTapLeaf(),
+	)
+	if err != nil {
+		return lfn.Err[asset.ScriptKey](err)
+	}
+
+	return lfn.Ok(asset.ScriptKey{
+		PubKey: localScriptTree.TaprootKey,
+		TweakedScriptKey: &asset.TweakedScriptKey{
+			RawKey: keychain.KeyDescriptor{
+				PubKey: localScriptTree.InternalKey,
+			},
+			Tweak: localScriptTree.TapscriptRoot,
+		},
+	})
+}
+
+// deriveCommitKeys derives the script keys for the local and remote party.
+func deriveCommitKeys(req lnwallet.ResolutionReq,
+) (*asset.ScriptKey, *asset.ScriptKey, error) {
+
+	localScriptTree, err := localCommitScriptKey(
+		req.KeyRing.ToLocalKey, req.KeyRing.RevocationKey,
+		req.CsvDelay,
+	).Unpack()
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to create local "+
+			"script key: %w", err)
+	}
+
+	remoteScriptTree, err := remoteCommitScriptKey(
+		req.KeyRing.ToRemoteKey,
+	).Unpack()
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to create remote "+
+			"script key: %w", err)
+	}
+
+	return &localScriptTree, &remoteScriptTree, nil
+}
+
+// importCommitScriptKeys imports the script keys for the commitment outputs
+// into the local addr book.
+func (a *AuxSweeper) importCommitScriptKeys(req lnwallet.ResolutionReq) error {
+
+	ctxb := context.Background()
```

Comment:
This line can be moved to just above `for _, key := range keysToImport {` below.